### PR TITLE
Removed Boost

### PR DESCRIPTION
**Pull request by esteve**
This PR removes Boost and uses the pure C++11 version of Asio instead.

### CODE REVIEW

**Comment by reza-ebrahimi** [comment]
Boost is not an extra dependency for autoware. Since all ROS messages are using boost, and autoware itself developing on top of ROS, What is the reason behind removing Boost?

**Comment by esteve** [comment]
@reza-ebrahimi I received a notification about your comment, but it does not show up here, that's weird. Anyway, responding to your comment:

> Boost is not an extra dependency for autoware. Since all ROS messages are using boost, and autoware itself developing on top of ROS, What is the reason behind removing Boost?

Only two packages in Autoware.Auto (where `udp_driver` and `serial_driver` are being used) use Boost, out of ~90. Also, ROS2 messages do not use Boost, in fact, one of our goals when we created ROS2 was to avoid using Boost as much as possible, as you can see, no core ROS2 packages have Boost as a dependency. There are several reasons for avoiding Boost, but mainly it's because most of the functionality found in Boost is already present in C++11/14/17 so there's no need for duplicating dependencies.

In this PR I also took the chance to make your implementation slightly more modern and avoid pre-C++11 constructs (e.g. no `typedef`, use lambdas where possible, use ROS2 logging mechanism instead of `iostream`, etc.)

**Comment by esteve** [comment]
@JWhitleyWork @flynneva @reza-ebrahimi this is now ready for review, please give it a look at it when you have a moment. Thanks.

File: io_context/io_context-extras.cmake [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Copyright 2018 Apex.AI, Inc.
```
Comment (flynneva) [inline io_context/io_context-extras.cmake:1]: I think this should be 2021

File: io_context/CMakeLists.txt [hunk 2]
```diff
@@ -52,12 +52,16 @@ if(BUILD_TESTING)
 
   ament_add_gtest(${TEST_UDP_DRIVER_EXE}
           test/${name}.cpp)
-  ament_target_dependencies(${TEST_UDP_DRIVER_EXE} std_msgs Boost)
+  ament_target_dependencies(${TEST_UDP_DRIVER_EXE} std_msgs "ASIO")
```
Comment (flynneva) [inline io_context/CMakeLists.txt:55]: @JWhitleyWork does the 'ament_auto' for testing find the std_msgs dependency automatically?

**Review by flynneva — APPROVED** [review]
Looks good to me. Although I don't fully understand the benefit of creating the 'FindASIO' cmake files separately, it seems to work out ok and probably is some standard somewhere 😅

**Review by JWhitleyWork — APPROVED** [review]
LGTM!

**Comment by esteve** [comment]
@flynneva @JWhitleyWork thanks for the reviews. I'll submit a followup PR to address the feedback and also a couple of cleanups I just thought of.

**Pull request merged**